src/ArrayReadStateMachine.cs

```diff
@@ -67,3 +67,23 @@ public ReadResult Read(ref Utf8JsonReader reader)
         }
     }
 }
+
+static class ArrayReadStateMachineReaderExtensions
+{
+    public static JsonReadError? TryRead<T>(this IJsonReader<T> reader,
```

Comment:
I did this in a hurry yesterday, but looking at it again, I think it's best to split this up into two methods (will reduce number of `out` and `ref` parameters) and also make it a non-extension.

Reply:
See bd51fc6e6b2cfd6328cedd90d9b77e95a76baee0 and feel free to resolve if it looks better to you.

Basically, the extension becomes an internal helper method of `ArrayReadStateMachineReader`, which reduces one `ref` parameter. Then I moved the splitting of the `JsonReadResult<>` into error and item into a proper method of `JsonReadResult<>`, where the `out` parameter shifts out. Now each can also be used independently.